rfcs/0000-changes-lifecyle.md

```diff
@@ -0,0 +1,318 @@
+
+```sway
+#![enable(some_experimental_feature)]
+```
+
+## Unstable book
+
+The idea of an unstable book is to be a repository of documentation, decisions, or even a devlog of the feature.
+Its unstructured nature is intentional and serves the purpose of unburdening developers to keep an update and formal
+documentation of a feature that will likely change.
+
+Ideally, each chapter will have a link to the GitHub issue, discussions, references and whatever else is necessary
+to allow stakeholders to give feedback on the new feature.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+This section contains the suggested guide on how to introduce changes to different parts of the compiler and associated tools.
+
+## `sway-features` crate
+
+A new crate named `sway-features` will be created and will contain **ALL** features and their metadata. The best suggestion is a macro to define features where enums, documentation, etc... will be generated.
+
+This macro also needs to generate code for the errors and warnings related to each error.
+
+Features will be parsed in this order:
+
+```rust
+let mut features = ExperimentalFeatures::default();
+features.parse_from_toml();
+features.parse_from_cli();
+features.parse_from_environment_variables();
+```
+
+This means that environment variables overwrite CLI arguments, which overwrite TOML configuration.
+
+If a feature requires or allows new configurations, these configurations should be optional, and as soon as confirmed the feature is enabled, it should verify if the required configuration is available.
+
+```
+> forc ... --experimental a_feature --a-feature-option 1
+```
+
+## Lexer and Parser
+
+To allow as user-friendly error messages as possible, in all possible cases, we want both the lexer and the parser to parse the new syntax even with the feature off.
+
+After that, the lexer and the parser will mark that an experimental feature was lexed/parsed; and a check will guarantee that no disabled experimental was parsed.
+
+The error message will have a message explaining that the feature is experimental and a Github link for more details on the stabilization lifecycle.
+
+```rust
+// always parse new syntax
+let new_syntax = parse_new_syntax();
+if ctx.experimental.is_disabled(Features::NEW_FEATURE) {
+   handler.emit_error(...);
+}
+```
+
+## Formatting
+
+Formatting should always format new syntax, even when the flag is off. To avoid breaking other parts of the code, or even worse, removing code.
+
+## LSP
+
+LSP can take advantage of specific error messages, and suggest user to enable the corresponding feature.
+
+## CST, AST and Typed Tree
+
+All trees must always support new features, which means that new nodes will always exist. Their specific behavior, desugaring, etc... will be gated by the experimental feature.
+
+More specifically, this means that variants should not be behind compiler flags.
+
+## Tests
```

Comment:
On the CI, are we going to run tests for all the features? What about the combinations of the features? We don't need to define that in this RFC, perhaps it's too early. But it will be an important practical question. One possibility would be to have a CI run with all experimental features off, and one with all on.

Reply:
What is going to define the configuration is the `test.*.toml`. The default `test.toml` will run the compiler with all experimental off (the same as no flags).

When a specific test needs a specific feature it will have a `test.feature_a.toml`, where that feature will be enabled. Now running our tests will run this specific test twice.

If we force a specific feature we can use the environment variable:

```
> FORC_EXPERIMENTAL=feature_a cargo r -p test --release
```